### PR TITLE
[IMP] account,*: Generic mixin for business computation

### PR DESCRIPTION
**Pull request by smetl**
Multiple things are duplicated in every business models and sometimes, done in a wrong way like taxes management.
For example:
    - price included taxes set on a product with fiscal position mapping
    - taxes using the round globally setting
    - default price_unit suggested on a line from product
Also, there is some feature that are badly managed like the custom seller on product that is currently working only with purchase but give different results on vendor bill.
In the future, we will need to unify all these behaviors to manage more easily the values to be set on EDI documents (only working on invoices right now).

task_id: 2654784

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Review by Feyensv — CHANGES_REQUESTED** [review]
As said before, this doesn't go far enough for me, the mixin should contain shared fields between models, but well, I can accept you don't want to go that far atm.

But after reading it more attentively, I also feel that it goes too far (yes, it's complex :D), you're giving birth to a "fourre-tout" you plan to integrate in multiple core models...

A big part of its logic makes no sense for the models you extend...

* You are making one mixin hosting both "Order" and "Order lines" concept, which makes no sense, at least for sale & purchase scopes (I don't care about the account module complexity, but do not add a bunch of meaningless methods on SO/SOL or PO/POL models)

* You're integrating custom module logic in this mixin (Cf the "fourre-tout" `_get_business_values`, with a bunch of optional things).  This means for me that some things should not be there, either delegated to models through specific hooks, or split into different mixins (product management, price management, tax management, ... we could even think of a display type mixin :D).

The idea to share and factorize some code between core models has been in my mind for a long time but I'm not convinced by the current state of this mixin/task/PR :thinking:.

File: addons/sale/models/sale_order.py [hunk 1]
```diff
@@ -15,7 +15,7 @@
 
 class SaleOrder(models.Model):
     _name = "sale.order"
-    _inherit = ['portal.mixin', 'mail.thread', 'mail.activity.mixin', 'utm.mixin']
+    _inherit = ['portal.mixin', 'mail.thread', 'mail.activity.mixin', 'utm.mixin', 'account.business.mixin']
```
[the remark this answers is not in the document]
Comment (legalsylvain) [reply addons/sale/models/sale_order.py:18]: :+100:

[the remark this answers is not in the document]
Comment (clementmbr) [reply addons/sale/models/sale_order.py:18]: :+100000:

[the remark this answers is not in the document]
Comment (lem8r) [reply addons/sale/models/sale_order.py:18]: Having different approach depending on object type seems like visitor pattern.

**Comment by fpodoo** [comment]
That PR looks like a bad idea to me. 

Instead of having each document (SO, PO, Invoice) that implements its own logic, we now have a "generic" method with hardcoded code like:
```
if business_vals['document_type'] == 'sale':
    ...
elif business_vals['document_type'] == 'purchase':
    ...
```
Not only it's harder to overwrite, but it also creates a more spagetthi code.

Eg: have a default value for sale.order.line that returns the name of the product with the right context (seems logic and simple), I now have to understand a mixin, inherits _get_business_values and hope the default is automated.


There are surely duplicated code that should be shared between SO & Invoice; but the right way to handle that is to call a common method: (e.g. the name_get of produt factors all ways of formatting a product name; no need for an extra layer).


I would have agreed if the PR comes with a high reduction of the number of lines of code. (like good rectoring should do). But this PR adds 400 LOC (excluding tests) for ~1000 lines changed (+40%), and an extra complexity. (to read/write code, you need to understand the extra layer of this mixin), which is a sign that it's not the right solution for the problem. (it does not help simplifying the code)

**Comment by fpodoo** [comment]
I believe the biggest refactoring in this area is to reimplement the onchange's that agregate multiple behaviors, into simpler compute method (without dependencies); that will unclutter the business logic and isolate methods according to their purpose, but keep these methods in their respective objects (SO, PO, Invoice).

By doing that, the code might be simpler and wont require a mixin. If some compute are too complex (e.g. name of a line, we might delegate the logic to other objects, like the name_get of products)

**Comment by legalsylvain** [comment]
Hi @fpodoo. Thanks for sharing your Point of view. 

I have a more nuanced opinion.

when i have discovered odoo (some long time ago (~10 years, OpenERP 5.0)) I found that it was a mess to have such similar "business objects" (purchase order / sale order / pos order) 
- with all logic duplicated
- with some messy differences. (line field is sometimes named ``line_ids``, sometimes ``order_line``, etc.. the same for ``uom`` fields)
- with some differences (in the business logic) that should not exist.

due to a lot of refactoring, the situation is better, but some points still remains.

I fully agree that if the refactor does'nt have a negative LOC balance, the refactor is wrong. However, having some light mixin that define at least the basic fields ``product_id``, ``uom_id``, etc... (for lines) and ``partner_id``, ``name``, etc.. (for main object) and some "basic onchange / compute / etc..." could be great.
Of course, if the mixin contains ``if model == "sale.order"`` the refactor failed. 

in a few world, maybe the current implementation is wrong. (I think it's hard to do, so it's normal to not having a great first draft it is not a critic against the authors), but the general idea should be keeped.

kind regards.

**Comment by smetl** [comment]
Don't say this PR is a bad idea after reading less than 10% of the code. Let's clarify some things. First of all, the purpose of this mixin is not only to refactor and clean the code but also to fix a lot of issues mainly with taxes (global rounding, fiscal position mapping price included taxes etc). So, you are comparing apples and oranges here because the current code is WRONG.

Then, I agree the mixin could be rewriting with static methods but think about the computation of taxes of all business lines. You will be forced to expressed it by passing a list of dictionaries as parameter (looking like the `_get_business_values` method) plus extra parameter defining how taxes should be computed or not. If all methods are static, it also means a lot of parameters to be set everywhere like:
`def _get_default_product_account(self, company, journal, fiscal_position, product, document_type):`
Do you really prefer calling such methods everywhere with a loooooot of parameters? I don't think so.

I also need to clarify the blabla about the number of lines. Let's take an example:
```python
    def _get_default_taxes(self):
        """ Get the default taxes.
        :return: An account.tax recordset.
        """
        business_vals = self._get_business_values()
        company = business_vals.get('company')
        fiscal_position = business_vals.get('fiscal_position')
        product = business_vals.get('product')
        account = business_vals.get('account')
        document_type = business_vals.get('document_type')

        if product and document_type == 'sale':
            taxes = product.taxes_id
        elif product and document_type == 'purchase':
            taxes = product.supplier_taxes_id
        else:
            taxes = self.env['account.tax']

        if company:
            taxes = taxes.filtered(lambda tax: tax.company_id == company)

        if not taxes and account:
            taxes = account.tax_ids

        if not taxes and company:
            if document_type == 'sale':
                taxes = company.account_sale_tax_id
            elif document_type == 'purchase':
                taxes = company.account_purchase_tax_id

        if taxes and fiscal_position:
            taxes = fiscal_position.map_tax(taxes)

        return taxes
```

This method is documented, pylint is happy with it, easy to understand. I can remove a lot of comments, remove all temporary variables to reduce the diff size but it's definitively not a good metric here because the current code is maybe shorter but WRONG. If you want the full analysis of the file:
```
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
Python                           1            125            148            445
-------------------------------------------------------------------------------
```

Here, 38% of the file is only there for lisibility and comments.

**Comment by legalsylvain** [comment]
Hi @smetl. 
Thanks for your answer. 
> Don't say this PR is a bad idea after reading less than 10% of the code

I don't know if you answer to me, but I must confess I have read only 10% of the diff ;-)
And sorry if my first comment was rough.

I still think that such refactor, introducing mixin is a very great idea, that is totally in the roadmap expressed last month by @fpodoo regarding the V16. (in a few words : "clean code").

regarding the number of lines, I think that considering [this remarks](https://github.com/odoo/odoo/pull/79564#discussion_r745780098) of @Feyensv could reduce code lines.

For exemple : 

### Current pseudo code
```
class AccountMoveLine(models.Model):
    _name = "account.move.line"

    product_id = fields.Many2one('product.product')
    product_uom_id = fields.Many2one('uom.uom')
    quantity = fields.Float()
    price_unit = fields.Float()
    tax_ids = fields.Many2many('account.tax')

class SaleOrderLine(models.Model):
    _name = 'sale.order.line'

    product_id = fields.Many2one('product.product')
    product_uom = fields.Many2one('uom.uom')
    product_uom_qty = fields.Float(string='Quantity')
    price_unit = fields.Float()
    tax_id = fields.Many2many('account.tax')

class PurchaseOrderLine(models.Model):
    _name = 'purchase.order.line'

     product_id = fields.Many2one('product.product')
     product_uom = fields.Many2one('uom.uom')
     product_qty = fields.Float()
     price_unit = fields.Float()
     taxes_id = fields.Many2many('account.tax')
```

### refactor proposal pseudo code
class AccountMoveLine(models.Model):
    _name = "account.move.line"
    _inherit = "account.business.line.mixin"

class SaleOrderLine(models.Model):
    _name = 'sale.order.line'
    _inherit = "account.business.line.mixin"

class PurchaseOrderLine(models.Model):
    _name = 'purchase.order.line'
    _inherit = "account.business.line.mixin"

class account.business.line.mixin(models.AbstractModel):
    _name = 'account.business.line.mixin'
    product_id = fields.Many2one('product.product')
    product_uom_id = fields.Many2one('uom.uom')
    quantity = fields.Float()
    price_unit = fields.Float()
    tax_ids = fields.Many2many('account.tax')
```

**Advantage** 
- Old code : 26 lines, new code 19 lines. (-26%). (I think it could be more, because I just took some fields, but more are interesting, like name, discount, price_subtotal, price_total, etc....)
- if I understand correctly ``_get_business_values`` doesn't need to return all the fields like 'product' (but some ones) : Just call ``self.product_id`` when required.
- all the same fields are named with the same name ! And that could be a great thing to announce in the next Odoo Experience ;-). I mean, for the time being, taxes fields are named ``tax_ids``, ``tax_id``, ``taxes_id``. But why ?!?

**temporary disadvantage**
- write migrations script to align all the diverging field names.

**limitation**
- some fields doesn't have all the same parameter (like domain), so overloading field should be required sometimes, but the main code is the same.

**Comment by randall-vx** [comment]
@hbto

**Comment by smetl** [comment]
@legalsylvain This mixin could also be applied on pos orders, expenses, subscriptions and I need it for another task for lines that doesn't have any product nor uom. Before starting a huge refactoring of anything, I want a global overview of all business objects.
Consider also some business models have a partner or a currency for every lines but it's not always the case (e.g. expenses, journal items,...).

**Comment by fpodoo** [comment]
Copy / pasting my discussion with Laurent as lots of people seem interested in this discussion.

En gros, je travaillerais en 2 temps: 
1/ On unclutter les onchange en redéveloppant les onchange en compute field (pas de mixin ni tools)
2/ Si trop de business logic est partagée entre les `compute`, on factorise juste celui là. (pas via une mixin mais dans la méthode appellée; product.name_get, tax.compute, ...

Mon feeling est que 80% de la simplification se fera à l'étape 1, et il n'y aura pas grand chose en 2.

Dit autrement, le problème n'est pas où mettre le code (dans une mixin, dans des tools, dans chaque objet), mais comment le coder, pour découpler les business logiques différentes. Il n'est pas normal d'avoir une méthode four-tout qui calcule à la fois le name, le prix, les taxes, la compte comptable, ... (product_id_change) car cela rend l'héritage galère.

Par exemple, la logique business pour obtenir le `unit_price` est répartie dans pleins de méthodes: default_get, product_id_change, qty_change, discount_change, ... Et il en est de même pour la plupart des champs de la ligne. Un méthode comme product_id_change a donc une partie de la logique business de chaque champ. (l'autre partie est dans les autres onchange) De plus cette logique, n'est que "client side", et ne peut être utilisée facilement dans le code qui manipule ces documents. (on a donc des appels dégueulasse avec pleins de paramètres à cause de l'API onchange)

Exemple de méthodes sur une ligne:
- product_id_change impacte name, account_id, tax_ids, unit price, qty, subtotal
- qty impacte tax_ids, unit price, subtotal
- name impacte account_id, ...

Ce serait plus simple de faire des computed fields, pour isoler la logique business de chaque champ dans une seule méthode:
- name = f(product_id)
- account_id = f(product_id, name)
- unit_price = f(product_id, qty)
- tax_ids = f(product_id)
-...

Bref, remplacer les onchange par des compute: ca rendra chaque fonction compute responsable d'un champ, au lieu de gros on_change qui font pleins de trucs, sur pleins de champs. Ca simplifiera les héritages car tout est découplé et les méthodes plus simples. Les relations entre ces champs est déléguée aux dependency graphs.

Une fois la business logique découplée, ce sera facile de factoriser une ou l'autre logique entre différents objets (ex: s'il y a trop de logique dans line.compute_tax_ids, on peut délèguer à une méthode sur product.product), mais je pense qu'il n'y aura même pas besoin. (ce sera minime en tous cas)

En remplaçant les onchange, on évitera aussi tous les appels explicite a onchange pour que la business logique s'applique. Tu pourrais faire dans le code ou les tests:

```
self.env['sale.order'].create({
    'partner_id': 3,
    'line_ids': [(0, 0, {'product_id': 5})]
})
```

et tout le SO sera correct avec les bonne fiscal position, prix, taxes, ... Ce qui doit simplifier du code un peu partout. Faire cela pour l'instant est une galère car tu dois invoquer des "onchange" à la main avec des paramètres.


L'optimisation de déplacer les champs dans une mixin pour éviter de les déclarer plusieurs fois est un détail que je ne ferais pas par clareté de code. On ne gagnerait que qq lignes, mais on perdrait en lisibilité et flexibilité (e.g. les name ou tips sur les champs ne sont pas nécessairement les même)

**Comment by clementmbr** [comment]
If I can add my 2 cents to @legalsylvain 's **advantage list** and as I understood one of the key point for Odoo growth is the new developers training : 

It's really painful as a beginner (and even after few years of Odoo) to be forced to stop and think "Is product's UoM called `product_uom` here or is it `product_uom_id` ?" when you're deeply focused on the code you're writing.

The basic DRY idea behind this PR will definitely help seducing and training new developers in the near future !

**Comment by smetl** [comment]
@clementmbr I agree with you but renaming the field could be done outside a mixin or in a second step. Imo, cleaning the current code should be done step by step and this PR is just adding the mixin without changing any business flow, field or anything. Also, I know very well the accounting code but not others modules so it's too risky for me to refactor everything in a single step

**Comment by legalsylvain** [comment]
> et tout le SO sera correct avec les bonne fiscal position, prix, taxes, ... Ce qui doit simplifier du code un peu partout. Faire cela pour l'instant est une galère car tu dois invoquer des "onchange" à la main avec des paramètres.

:+1: If I understand correctly, this could make obsolete current oca tools ``onchange_helper``. https://github.com/OCA/server-tools/tree/14.0/onchange_helper
CC : @guewen, @florian-dacosta, ...

> @legalsylvain This mixin could also be applied on pos orders, expenses, subscriptions

I don't follow you on that point. A "business" mixin should be applied to models that shares similar "business" concept. For me ``sale.order.line`` and ``purchase.order.line`` have a lot of common thing. but ``sale.order.line`` and ``subscription`` not. 
But maybe I'm wrong. 

thanks for this interesting discussion anyway. And thanks @fpodoo to share your thoughts.

**Comment by fpodoo** [comment]
> +1 If I understand correctly, this could make obsolete current oca tools onchange_helper.
Yes, our goal is to remove onchanges and replace them by compute methods.

Here is a snippet from the tests in sale:

```
cls.sale_order = SaleOrder.create({
       'partner_id': cls.partner_a.id,
       'partner_invoice_id': cls.partner_a.id,
       'partner_shipping_id': cls.partner_a.id,
       'pricelist_id': cls.company_data['default_pricelist'].id,
})
cls.sol_product_order = cls.env['sale.order.line'].create({
       'name': cls.company_data['product_order_no'].name,
       'product_id': cls.company_data['product_order_no'].id,
       'product_uom_qty': 2,
       'product_uom': cls.company_data['product_order_no'].uom_id.id,
       'price_unit': cls.company_data['product_order_no'].list_price,
       'order_id': cls.sale_order.id,
       'tax_id': False,
   })
```

If the logic would be in fields `compute readonly="False"`instead of `onchange`, you could write for the exact same effect:
```
cls.sale_order = SaleOrder.create({
       'partner_id': cls.partner_a.id,
})
cls.sol_product_order = cls.env['sale.order.line'].create({
       'product_id': cls.company_data['product_order_no'].id,
       'product_uom_qty': 2,
       'order_id': cls.sale_order.id,       'order_id': cls.sale_order.id,

})
```

In this example, the business logic was hard coded (the partner_invoice_id is the address of partner_id), but in the real code, it's even worse; you have to call plenty of onchanges with lots of parameters to make the business logic apply. We would save hundreds of lines of code by this refactoring. (much moer than trying to share a few fields on a mixin)

**Comment by Feyensv** [comment]
@fpodoo FYI, the onchange -> compute work for the sale scope is (once again) in progress, in #79093.
Cannot guarantee this will be merged for next saas, but should be for v16 :crossed_fingers:.

**Comment by legalsylvain** [comment]
> I bet nobody can say exactly what a product_id_change does as it impacts lots of fields, and it depends on modules instead

:+1: 

> @fpodoo FYI, the onchange -> compute work for the sale scope is (once again) in progress, in #79093.

Nice.

image
``<troll-mode>``
The balance of LOC is not good... Is it a good refactor ?
``</troll-mode>``

**Comment by hbto** [comment]
> @hbto

@randall-vx reading it.

Regards.

**Comment by fpodoo** [comment]
> if you change the uom_id of the product, it will change the uom_id of all the existing sale.order.line. (Or did I missed something in the recent related refator ?

No, because the depends is "product_id" and not "product_id.uom_id".

**Pull request closed without merging**